File: my-folder/0040-combination-sum-ii/solution.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>>res;
        vector<int>combi;
        sort(candidates.begin(),candidates.end());
        for(int x:candidates)
            cout<<x<<" ";
        helper(candidates, target, res, combi, 0);
        return res;
    }
    
    void helper(vector<int>& candidates, int target, 
                vector<vector<int>>& res, vector<int>& combi, int index)
    {
        if(target<=0)
        {
            res.push_back(combi);
            return;
        }
        
        for(int i=index;i<candidates.size()&&target>=candidates[i];i++)
        {
            if(i==index or candidates[i]!=candidates[i-1])
            {
                combi.push_back(candidates[i]);
                helper(candidates, target-candidates[i], res, combi, i+1);
                combi.pop_back();
            }
        }
    }
};

File: my-folder/0040-combination-sum-ii/solution.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        res=new ArrayList<>();
        Arrays.sort(candidates);
        solve(0, candidates, target, new ArrayList<Integer>());
        return res;
        
    }

    public void solve(int index, int[] candidates, int target, List<Integer> curr){
        if(target==0){
            res.add(new ArrayList<Integer>(curr));
            return;
        }
        else{
            for(int i=index;i<candidates.length;i++){
                if(i>index && candidates[i]==candidates[i-1])
                    continue;
                if(candidates[i]>target)
                    break;
                curr.add(candidates[i]);
                solve(i+1,candidates, target-candidates[i], curr);
                curr.remove(curr.size()-1);                
            }
        }
    }
}
